presto-main/src/main/java/com/facebook/presto/server/QueryResource.java

```diff
@@ -84,6 +87,14 @@ public void cancelQuery(@PathParam("queryId") QueryId queryId)
         queryManager.cancelQuery(queryId);
     }
 
+    @PUT
+    @Path("{queryId}/kill")
+    public void killQuery(@PathParam("queryId") QueryId queryId, String reason)
```

Comment:
Is this idempotent? `PUT` should be idempotent. What happens when we call `failQuery` for a query that's already failed?

Reply:
This is idempotent, but we are not returning an error if the call fails because the query is already done, or because the query is not found. I'm not sure we care about either case, but it's worth considering.

Reply:
If we make it fail because the query is already done, then we need to make it idempotent by not failing if the query is already killed and has the same reason.

Similar considerations apply to the `kill_query` procedure.

Reply:
Currently it is idempotent because we are not doing anything about the return value from stateMachine.transitionToFailed(cause) in fail(cause), which return false if the current state is done. In such case, stateMachine.transitionToFailed() doesn't do anything but just logging it. Also the kill_query procedure doesn't return a true/false and is idempotent. What might be the cases to make them non-idempotent?
Another question is shall we make the kill_query procedure to call the failQuery(reason) instead of cancelQuery()? If it's a publicly available function maybe we should just let it follow the old way since it's for the users to cancel a query with error "USER_CANCELED".